### DISCUSSION TITLE
Default arguments that aren't compile-time constants

### DISCUSSION THREAD

**Discussion by jaigak** [dotnet/csharplang#5489] [Language Ideas] [answered]
Today in C#, default arguments for methods must have a compile-time constant as its value. However, in many cases, it may be a valid use case for having a value resolved at runtime. The solution is to either use an overload that fills in the default arguments or set the default argument to a value that indicates it's not initialized (eg: null for reference types) and fill it in within the method if the value matches the uninitialized value. Both of which don't clearly express to the user that there's a default argument involved unless you specify it in comments or documentation. Many other languages such as C++ support having a default argument that's not just compile-time constant value but resolved at runtime. Here's an example involving DateTime:
```csharp
SomeType GetFromDateTime(DateTime dateTime = DateTime.Now)
```
Today this wouldn't compile as DateTime.Now isn't a compile-time constant.
You might write this instead:
```csharp
SomeType GetFromDateTime() => GetFromDateTime(DateTime.Now);

SomeType GetFromDateTime(DateTime dateTime);
```
Except that now the user doesn't know it's actually doing the exact same thing but the time is assumed to be now.
Or this:
```csharp
SomeType GetFromDateTime(DateTime? dateTime)
{
    if (dateTime == null) dateTime == DateTime.Now;
}

```
Whereas in C++, you could just write this:
```cpp
SomeType GetFromDateTime(DateTime dateTime = clock::now());
```

I do understand that the default argument information is stored in metadata and is independent of language so consider adding support for default arguments resolved at runtime only if that expression can be expressed in any .NET language (eg: property access) and requires the same access level as the method itself (so the user can see and use it).

* The method GetFromDateTime doesn't make much sense to have a default argument however this is just an example.

**Comment by HaloFour** [comment]
IMO this is a case where overloads are more appropriate than default parameters.  They can be more verbose, but you have more control over the API surface declaring them manually.  I don't believe that default parameters could generate overloads because as soon as you have more than one you end up with combinatorial explosion of possibilities and the likelihood that the signatures of two variants will be identical.

As for relying on sentinel values, that is probably a bit easier but it also only really works for reference types and only where `null` is ever expected to be a legit value to be passed.  Passing the sentinel value directly may have unexpected results.

Scala has an interesting approach here.  If a Scala method accepts a parameter value the compiler emits a special public method that the consumer of that method can call to get that default value.  That enables the default value to be of any expression:

```scala
// declaration
def getSomeDateTime(dateTime: DateTime = DateTime.now()) = ??? 

// invocation
val result = getSomeDateTime()

// lowered to
val result = getSomeDateTime(getSomeDateTime$default());
```

The problem with this approach being that the compiler needs to generate a unique but predictable name for each of these methods, and as they are a part of the public surface of that API any deviation can break existing consumers.

**Reply by jaigak** [reply]
But then you don't know what the value of the default parameter is.

**Reply by HaloFour** [reply]
The consumer wouldn't know the value of the default parameter in the case of generated overloads or sentinel values either.

**Reply by jaigak** [reply]
Well if you used a default argument, IntelliSense will show the value.

**Reply by HaloFour** [reply]
IntelliSense can do that because it can interpret the default parameter metadata attached to the method.  It can do the same if another approach was used.  IntelliJ has no problem displaying the expression of a default value in Scala:

![image](https://user-images.githubusercontent.com/10102290/144253319-a2bcff56-3865-4b4a-8b37-f488249f569b.png)

**Comment by jnm2** [comment, marked as answer]
Duplicate of https://github.com/dotnet/csharplang/discussions/1702 and https://github.com/dotnet/csharplang/discussions/3947
